In [11]:
import pandas as pd
import numpy as np

# Load dataset

In [10]:
df = pd.read_csv("car_dataset_.csv")
df

Car_ID    Brand        Model    Year Fuel_Type Transmission  \
0        1.0     Ford  Convertible  2019.0       NaN       Manual   
1        2.0    Honda          SUV  2015.0    Diesel    Automatic   
2        3.0     Audi          SUV  2001.0  Electric    Automatic   
3        4.0  Hyundai        Sedan  2016.0    Hybrid    Automatic   
4        5.0      NaN        Truck  2021.0    Hybrid    Automatic   
...      ...      ...          ...     ...       ...          ...   
8995  8996.0   Toyota        Truck  2001.0    Petrol    Automatic   
8996  8997.0    Honda    Hatchback  2003.0    Hybrid       Manual   
8997  8998.0     Audi          SUV  2011.0    Diesel       Manual   
8998     NaN      NaN        Truck     NaN    Petrol    Automatic   
8999  9000.0     Audi  Convertible  2006.0    Diesel          NaN   

      Engine_Size  Horsepower  Mileage    Price Owner_Type   Color  
0             NaN       267.0     19.0  62151.0      First   Black  
1             3.1         NaN     20.4  25861.0      First   White  
2             2.6       213.0      9.5  61772.0      Third    Grey  
3             4.3       104.0     26.1  16354.0     Second     Red  
4             4.4         NaN      6.5  88664.0      Third    Grey  
...           ...         ...      ...      ...        ...     ...  
8995          3.8       181.0     12.3  52730.0     Second  Silver  
8996          3.1       469.0      7.8  60988.0      First  Silver  
8997          4.2       255.0     19.5  74528.0      Third   Black  
8998          3.3       109.0     28.5  98910.0      First     NaN  
8999          4.1        77.0     24.9  58484.0     Second    Grey  

[9000 rows x 12 columns]

# Obsersvation
- Dataset has been loaded successfully using pd.read_csv()

# Understanding the Structure of the Dataset¶

In [12]:
df.shape

(9000, 12)

# Observation
- The dataset contains 9000 rows and 12 columns representing various car-related features.

# Previewing the first few rows


In [13]:
df.head()

Car_ID    Brand        Model    Year Fuel_Type Transmission  Engine_Size  \
0     1.0     Ford  Convertible  2019.0       NaN       Manual          NaN   
1     2.0    Honda          SUV  2015.0    Diesel    Automatic          3.1   
2     3.0     Audi          SUV  2001.0  Electric    Automatic          2.6   
3     4.0  Hyundai        Sedan  2016.0    Hybrid    Automatic          4.3   
4     5.0      NaN        Truck  2021.0    Hybrid    Automatic          4.4   

   Horsepower  Mileage    Price Owner_Type  Color  
0       267.0     19.0  62151.0      First  Black  
1         NaN     20.4  25861.0      First  White  
2       213.0      9.5  61772.0      Third   Grey  
3       104.0     26.1  16354.0     Second    Red  
4         NaN      6.5  88664.0      Third   Grey

# Obsersavtion
- The preview shows the first few rows of the data and helps us check if the file loaded correctly.

# Dataset Structure and Data Types

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Car_ID        8100 non-null   float64
 1   Brand         8100 non-null   object 
 2   Model         8100 non-null   object 
 3   Year          8100 non-null   float64
 4   Fuel_Type     8100 non-null   object 
 5   Transmission  8100 non-null   object 
 6   Engine_Size   8100 non-null   float64
 7   Horsepower    8100 non-null   float64
 8   Mileage       8100 non-null   float64
 9   Price         8100 non-null   float64
 10  Owner_Type    8100 non-null   object 
 11  Color         8100 non-null   object 
dtypes: float64(6), object(6)
memory usage: 843.9+ KB


I have 900 missing values per column (9000 total rows, but only 8100 valid in each column).

Data types are mixed:

float64 → numerical (Car_ID, Year, Engine_Size, Horsepower, Mileage, Price).

object → categorical (Brand, Model, Fuel_Type, Transmission, Owner_Type, Color).

Memory usage: ~844 KB (lightweight dataset).

This dataset looks like a car sales/records dataset with features about car specs + sales details.

# Statistical Overview of Numerical Data

In [25]:
df.describe()

Car_ID         Year  Engine_Size   Horsepower      Mileage  \
count  8100.000000  8100.000000  8100.000000  8100.000000  8100.000000   
mean   4484.535926  2008.524321     2.997037   278.152469    17.471049   
std    2599.169860     8.087283     1.147651   126.697237     7.218944   
min       1.000000  1995.000000     1.000000    60.000000     5.000000   
25%    2233.750000  2002.000000     2.000000   168.000000    11.200000   
50%    4479.500000  2008.000000     3.000000   278.000000    17.500000   
75%    6740.250000  2016.000000     4.000000   388.000000    23.700000   
max    9000.000000  2022.000000     5.000000   499.000000    30.000000   

              Price  
count   8100.000000  
mean   51033.630617  
std    28498.359118  
min     2005.000000  
25%    26248.000000  
50%    51313.500000  
75%    75768.500000  
max    99998.000000

# Observation
The dataset looks like a mix of economy, mid-range, and luxury cars.

Most cars are 2000–2015 models, with engine size around 3.0L, horsepower 278 HP, mileage 17–18 km/l, and price ₹50k approx.

Outliers exist at the lower (very cheap, low-power cars) and upper end (luxury, high-power cars).

# Null value count Across All columns

In [26]:
df.isnull().sum()

Car_ID          900
Brand           900
Model           900
Year            900
Fuel_Type       900
Transmission    900
Engine_Size     900
Horsepower      900
Mileage         900
Price           900
Owner_Type      900
Color           900
dtype: int64

# Percentage of Missing Values¶

In [30]:
round(df.isnull().sum()/len(df)*100)

Car_ID          10.0
Brand           10.0
Model           10.0
Year            10.0
Fuel_Type       10.0
Transmission    10.0
Engine_Size     10.0
Horsepower      10.0
Mileage         10.0
Price           10.0
Owner_Type      10.0
Color           10.0
dtype: float64

# Observation:
- Every column (Car_ID, Brand, Model, Year, Fuel_Type, Transmission, Engine_Size, Horsepower, Mileage, Price, Owner_Type, Color) has 10% missing values.

This matches the dataset we designed (we randomly introduced ~10% NaNs).

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Car_ID        8100 non-null   float64
 1   Brand         8100 non-null   object 
 2   Model         8100 non-null   object 
 3   Year          8100 non-null   float64
 4   Fuel_Type     8100 non-null   object 
 5   Transmission  8100 non-null   object 
 6   Engine_Size   8100 non-null   float64
 7   Horsepower    8100 non-null   float64
 8   Mileage       8100 non-null   float64
 9   Price         8100 non-null   float64
 10  Owner_Type    8100 non-null   object 
 11  Color         8100 non-null   object 
dtypes: float64(6), object(6)
memory usage: 843.9+ KB


# Value count summary 

### Car_ID
- Dataset includes various of car ID , The name of the Car_ID
- example : 1.0,2.0,3.0.


### Brand
- represents the sector in which have different types of Brands
- Example : Ford, Honda, Audi, Hyundai, Kia, BMW, Toyota

### Model
- In This have different types of Model
 - Example:Convertible, SUV, Sedan, Truck, Hatchback, Coupe.

### Year 
- Manufacturing year of the car

### Fuel_Type 
- Petrol, Diesel, Electric, Hybrid

### Transmission 
- Manual, Automatic

### Engine_Size 
- Engine capacity (e.g., 2.0L, 3.5L)

### Horsepower 
- Power of the engine

### Mileage 
-Fuel efficiency (km/l)

### Price 
- Cost of the car

### Owner_Type 
- First owner, Second owner, etc.

### Color 
- Car color

In [38]:
df['Car_ID'].value_counts()

Car_ID
1.0       1
5960.0    1
5975.0    1
5974.0    1
5972.0    1
         ..
2985.0    1
2984.0    1
2983.0    1
2982.0    1
9000.0    1
Name: count, Length: 8100, dtype: int64

In [39]:
df["Year"].value_counts()

Year
2007.0    325
2004.0    317
2010.0    316
2021.0    314
2006.0    304
2022.0    300
2009.0    300
2001.0    297
1996.0    295
2020.0    294
2015.0    294
2000.0    289
1999.0    289
2018.0    287
1995.0    286
2019.0    283
2005.0    282
2003.0    282
1998.0    281
1997.0    281
2013.0    281
2016.0    281
2012.0    276
2017.0    276
2011.0    273
2008.0    269
2014.0    265
2002.0    263
Name: count, dtype: int64

## observation:
- In this the datset its shows year of buying cars and each year how many cars buying

In [40]:
df["Engine_Size"].value_counts()

Engine_Size
3.4    230
3.3    220
3.5    219
3.7    219
4.7    218
1.3    217
3.8    217
1.1    216
3.2    216
2.5    214
4.2    214
2.2    213
1.5    213
2.7    212
2.4    210
1.9    209
1.8    208
2.3    208
4.8    206
2.0    204
4.3    203
4.0    203
4.9    201
3.1    200
2.6    199
2.9    197
2.1    197
3.9    196
4.4    195
1.6    194
4.1    191
1.2    188
4.5    185
1.7    185
3.0    185
2.8    182
3.6    182
1.4    178
4.6    169
1.0     94
5.0     93
Name: count, dtype: int64

## Observation:
- In this data set contains the size of the engine and its valuecounts 

In [43]:
df["Horsepower"].value_counts()

Horsepower
267.0    31
336.0    31
141.0    30
388.0    30
279.0    30
         ..
459.0     9
468.0     9
313.0     9
91.0      9
418.0     7
Name: count, Length: 440, dtype: int64

## Observation:
- In this dataset shows about the horse power of the value counts

In [44]:
df["Mileage"].value_counts()

Mileage
23.5    46
22.5    44
8.8     44
26.2    44
16.9    43
        ..
26.6    21
22.9    20
14.8    20
5.7     19
5.0     14
Name: count, Length: 251, dtype: int64

## Observation:
- Its shows the mileage of the car in the data set

In [45]:
df['Price'].value_counts()

Price
41693.0    3
80558.0    3
69252.0    3
94577.0    3
6337.0     3
          ..
25298.0    1
16683.0    1
71962.0    1
28400.0    1
58484.0    1
Name: count, Length: 7801, dtype: int64

## Observation:
- its showns the price of the car and which car prices most frequently price

# Real Numerical Columns
- Car_ID → numeric ID (but it’s an identifier, not useful for analysis).

- Brand → categorical (Ford, Honda, etc.).

- Model → categorical (SUV, Sedan, etc.).

- Year → numeric (car manufacturing year).

- Fuel_Type → categorical (Petrol, Diesel, Electric, Hybrid).

Transmission → categorical (Manual, Automatic).

- Engine_Size → numeric (in Liters).

- Horsepower → numeric (power of engine).

- Mileage → numeric (fuel efficiency, km/l).

- Price → numeric (car cost).

- Owner_Type → categorical (First, Second, Third).

- Color → categorical (Black, White, Red, etc.).

# Discrete Numerical Columns
- Discrete numeric (treat as integers):
- Year (e.g., 2016, 2021)
- Horsepower (often recorded as whole numbers)
- Price (recorded as whole numbers)
- Car_ID (identifier — discrete, but usually exclude from analysis)

# Checking for Duplicates¶

In [46]:
df.duplicated().sum()

0

## Observation:
- As seen there are no duplicates found in the dataset

## Detecting Outliers
- The IQR (Interquartile Range) method was used to detect outliers in all numerical columns as part of the non-visual analysis.

### steps:
- Calculate Q1 (25th percentile) and Q3 (75th percentile)
- Compute IQR = Q3 - Q1
- Define lower and upper bounds:
- Lower bound = Q1 − 1.5 × IQR
- Upper bound = Q3 + 1.5 × IQR
 -Any value outside this range is considered an outlier

In [47]:
num_cols = df.select_dtypes(include=['float64', 'int64']).columns

In [ ]:
outliers = {}

In [48]:
outliers = {}
for col in num_cols:
    if df[col].isnull().all():
        continue  # Skip the columns with all missing values
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    outlier_values = df[(df[col] < lower) | (df[col] > upper)][col]
    if not outlier_values.empty:
        outliers[col] = outlier_values.tolist()

In [49]:
if outliers:
    for col, vals in outliers.items():
        print(f"\nOutliers in column '{col}':\n{vals}")
else:
      print("No outliers found in the numerical columns.")

No outliers found in the numerical columns.


## Observation
- Using the IQR method, outliers were detected in the Projected Openings (ZERO) column.

# Phase 2